### PR TITLE
Implement DeepDrug model

### LINKED ISSUE #14: Add the DeepDrug Model
Dear @kajocina, 

- Please read the paper first. It is [here](https://www.biorxiv.org/content/biorxiv/early/2020/11/10/2020.11.09.375626.full.pdf).
- There is also code-release with the paper [here](https://github.com/wanwenzeng/deepdrug).
- After that read the contributing guidelines.
-   If there is an existing open-source version of the model please take a look.
-   ChemicalX is built on top of [PyTorch 1.10.](https://pytorch.org/) and [torchdrug](https://torchdrug.ai/).
- A similar model is `EPGCNDS` which uses`GraphConvolutions`  to generate drug representations. Take a look at the layer definition [here](https://github.com/AstraZeneca/chemicalx/blob/main/chemicalx/models/epgcnds.py). You should use the layers from `torchdrug` not the models.
- The library heavily builds on top on [torchdrug](https://torchdrug.ai/) and molecules in batches are PackedGraphs.
- There is already a model class under `./chemicalx/models/`
- Context features, drug level features, and labels are all FloatTensors.
- Look at the examples and tests under `./examples/` and `./tests/`.
- Add auxiliary layers as you see fit - please document these, add tests, and add these layers to the main readme.md if needed.
- Add typing to the initialization and forward pass.
- Non-data-dependent hyper-ammeters should have default values. 
- Please add tests under `./tests/` and make sure that your model/layer is tested with real data.
- Write an example under `./examples/`. What is the AUC on the test set? Is it reasonable?

### PR DESCRIPTION
**Pull request by kajocina**
Closes #14

Added the DeepDrug model, trying to copy as much as possible from their approach. The paper didn't use any context features, but I implemented the model so that it could be used in both ways. I ran the model with and without context feats on DrugCombDB with around 0.76 AUROC (no context features led to a minor drop in AUROC).

Provided an example that runs the model and a unit test which tests both the context and no-context modes of this model.
 
- [X] Unit tests provided for these changes
- [X] Documentation and docstrings added for these changes

### CODE REVIEW

**Comment by cthoyt** [comment]
@kajocina please pass CI before we review this PR. You can either read the results in the GitHub browser or run `tox` locally to see all issues that need to be fixed

File: chemicalx/models/deepdrug.py [hunk 1]
```diff
@@ -1,14 +1,102 @@
...
     "DeepDrug",
 ]
 
 
-class DeepDrug(UnimplementedModel):
+class DeepDrug(Model):
     """An implementation of the DeepDrug model.
 
-    .. seealso:: https://github.com/AstraZeneca/chemicalx/issues/14
+    .. based on:: https://www.biorxiv.org/content/biorxiv/early/2020/11/10/2020.11.09.375626.full.pdf
```
Comment (cthoyt) [inline chemicalx/models/deepdrug.py:21]: please put an actual citation

Comment (cthoyt) [reply chemicalx/models/deepdrug.py:21]: `seealso` is also a special directive in RST, not sure why you changed it to `based on` which causes a parse error

File: chemicalx/models/deepdrug.py [hunk 2]
```diff
@@ -1,14 +1,102 @@
...
+        :param dropout_rate: Dropout rate on the final fully-connected layer.
+        """
+        super(DeepDrug, self).__init__()
+        self.num_gcn_layers = num_gcn_layers
+        self.gcn_layer_hidden_size = gcn_layer_hidden_size
+        self.graph_convolution_first = GraphConv(molecule_channels, self.gcn_layer_hidden_size, batch_norm=True)
+
+        # add remaining GCN layers
+        self.gcn_layer_list = torch.nn.ModuleList()
+        for i in range(num_gcn_layers - 1):
```
Comment (cthoyt) [inline chemicalx/models/deepdrug.py:51]: `i` is not used in this loop

```suggestion
         for _ in range(num_gcn_layers - 1):
```

File: chemicalx/models/deepdrug.py [hunk 3]
```diff
@@ -1,14 +1,102 @@
...
+
+    def forward(
+        self, context_features: Optional[torch.FloatTensor], molecules_left: PackedGraph, molecules_right: PackedGraph
+    ) -> torch.FloatTensor:
+
+        features_left = self.graph_convolution_first(molecules_left, molecules_left.data_dict["node_feature"])
+        features_right = self.graph_convolution_first(molecules_right, molecules_right.data_dict["node_feature"])
+
+        # run remaining GCN layers
+        for i in range(self.num_gcn_layers - 1):
```
Comment (cthoyt) [inline chemicalx/models/deepdrug.py:85]:
```suggestion
        for layer_i in self.gcn_layer_list:
```

**Comment by cthoyt** [comment]
I would suggest implementing this class only to have the functionality as originally described, then to consider subclassing it to have your additional functionality (adding in the context features may actually make it the same as one of the other models)

File: chemicalx/models/deepdrug.py [hunk 4]
```diff
@@ -1,14 +1,102 @@
...
+            layer_i = self.gcn_layer_list[i]
+            features_left = layer_i(molecules_left, features_left)
+            features_right = layer_i(molecules_right, features_right)
+
+        features_left = self.max_readout(molecules_left, features_left)
+        features_right = self.max_readout(molecules_right, features_right)
```
Comment (cthoyt) [inline chemicalx/models/deepdrug.py:91]: the same exact code is applied to molecules_left and molecules_right. This is a great chance to make a helper function to deduplicate code

**Comment by kajocina** [comment]
> @kajocina please pass CI before we review this PR. You can either read the results in the GitHub browser or run `tox` locally to see all issues that need to be fixed

@cthoyt I am getting CI errors from code in our tox dependencies so I can't really fix those, e.g.: 

```
lint/bin/activate_this.py:28: error: "str" has no attribute "decode"; maybe "encode"?
lint/bin/activate_this.py:31: error: Module has no attribute "real_prefix"
Found 2 errors in 1 file (checked 39 source files)
```
I did fix tox issues related to my commits now.

**Comment by cthoyt** [comment]
@kajocina you can try running tox where it automatically rebuilds the virtualenvs with `tox -r`. If that's not going well you can even consider reinstalling tox itself. However, the CI run on GitHub actions still shows issues in the lint and documentation tasks. Please look into those

**Comment by kajocina** [comment]
@cthoyt thanks for the tip, rebuilt tox and works ok, local CI was green now. I added the wrapper as you suggested, but it didnt really decrease the number of lines of code in total, only the number of assignments performed.

**Comment by cthoyt** [comment]
@kajocina the goal was to make the model more legibile. I will demonstrate by pushing to your branch.

**Review by cthoyt — APPROVED** [review]

**Review by benedekrozemberczki — APPROVED** [review]
The _helper functions are great.

**Comment by cthoyt** [comment]
Now that we have all of the implementations, I can see some interesting abstractions we can provide!
